tour_model_eval/confirmed_trips_eval_bins_clusters.py

```diff
@@ -9,65 +9,252 @@
+    non_empty_trips = [t for t in trips if t["data"]["user_input"] != {}]
+    non_empty_trips_df = pd.DataFrame(t["data"]["user_input"] for t in non_empty_trips)
+    valid_trips_df = non_empty_trips_df.dropna(axis=0, how='any', thresh=None, subset=None, inplace=False)
+    valid_trips_idx_ls = valid_trips_df.index.tolist()
+    valid_trips = [non_empty_trips[i]for i in valid_trips_idx_ls]
+
+    # similarity codes can filter out trips that are points in valid_trips
+    sim = similarity.similarity(valid_trips, radius)
+    filter_trips = sim.data
+    return filter_trips,sim,trips
+
+
+# to determine if the user is valid:
+# valid user should have >= 10 trips for further analysis and the proportion of filter_trips is >=50%
+def valid_user(filter_trips,trips):
+    valid = False
+    if len(filter_trips) >= 10 and len(filter_trips) / len(trips) >= 0.5:
+        valid = True
+    return valid
+
+
+# to map the user labels
+# - user_input_df: pass in original user input dataframe, return changed user input dataframe
+# - sp2en: change Spanish to English
+# - cvt_pur_mo: convert purposes and replaced mode
+def map_labels(user_input_df,sp2en,cvt_pur_mo):
+    # Spanish words to English
+    span_eng_dict = {'revisado_bike': 'test ride with bike', 'placas_de carro': 'car plates', 'aseguranza': 'insurance',
+                     'iglesia': 'church', 'curso': 'course',
+                     'mi_hija recién aliviada': 'my daughter just had a new baby',
+                     'servicio_comunitario': 'community service', 'pago_de aseguranza': 'insurance payment',
+                     'grupo_comunitario': 'community group', 'caminata_comunitaria': 'community walk'}
+
+    # Convert purpose
+    map_pur_dict = {'course': 'school', 'work_- lunch break': 'lunch_break', 'on_the way home': 'home',
+                    'insurance_payment': 'insurance'}
+
+    if sp2en:
+        # change language
+        user_input_df = user_input_df.replace(span_eng_dict)
+    elif cvt_pur_mo:
+        # change language first
+        user_input_df = user_input_df.replace(span_eng_dict)
+        # convert purpose
+        user_input_df = user_input_df.replace(map_pur_dict)
+        # convert mode
+        for a in range(len(user_input_df)):
+            if user_input_df.iloc[a]["replaced_mode"] == "same_mode":
+                # to see which row will be converted
+                logging.debug("The following rows will be changed: %s", user_input_df.iloc[a])
+                user_input_df.iloc[a]["replaced_mode"] = user_input_df.iloc[a]['mode_confirm']
+    return user_input_df
+
+
+# check if the user is valid
+# append NaN to the score lists when the user invalid
+def valid_user_check(filter_trips,trips,homo_score,comp_score,v_score):
+    if not valid_user(filter_trips, trips):
+        homo_score.append(NaN)
+        comp_score.append(NaN)
+        v_score.append(NaN)
+        skip = True
+    else:
+        skip = False
+    return homo_score,comp_score,v_score,skip
```

Comment:
I'm not going to insist on this since you are planning to refactor anyway, but you don't need to have `homo_score`, `comp_score` and `v_score` as both input and output. You can create a class that encapsulates all the scores and just pass it in directly.

Also, I really don't think you need to have append code in here - you can just have this function return valid or invalid. If invalid, append the values and continue; if valid, proceed with computation.

Reply:
Please file an issue to clean this up later.